## Writing Your Own Functions

So far we have gained an understanding of how to use other people's functions, writing our own loops and putting logical statements within them. 

The next step is to learn how to write your own functions which can then be applied to anything we put into them. In simple ones it is no different than what we have already done just reusable.

In [10]:
def printing(x):
    print(x)

In [12]:
printing("Hello")

Hello


This is the simplest and most redundant function ever to be used but it is to demonstrate the principal.

In essence you use def as a command to define your function. You put the function name and then the inputs your function has into the brackets:

def FUNCTIONNAME (INPUT): 
    Whatever you want to do to the input
    
Let's try something a little bit more realistic.

In [14]:
def studentinfo(Name,Stream,College):
    print("Student's name is",Name)
    print("They are in the", Stream, "Stream")
    print("They are studying at", College, "College")

In [16]:
studentinfo("Emily","BEEP","Brasenose")

Student's name is Emily
They are in the BEEP Stream
They are studying at Brasenose College


Try out a bunch of student names, streams and colleges. The point is that once you have defined the function you can put whatever you want into it and it will do the same thing. It is useful if you want to reuse the same thing and don't want to write out the code every time.

Generally speaking in data science / scientific computing we will be creating mathematical or analytical functions. 

Let's try one.

In [17]:
import numpy as np

In [34]:
def multbyrand(x):
    z = np.random.randint(50)
    y = x * z
    return(y)

In [35]:
multbyrand(1)

16

In [38]:
multbyrand(15)

675

This is another simple function that multiplies whatever number we put into it by a random number between 0 and 50. 

Next thing we need to do is to start incorporating our loops and if statements into our functions.

In [53]:
def extractindex(x,array):
    for i in range(0,len(array)):
        z = array[i]
        if i == x:
            return(z)

In [45]:
n = np.arange(23,100,1)

In [54]:
extractindex(10,n)

33

What we have done is create a function that outputs the value in any array at the index that we specify. We can also save the results of our function as a variable.

In [56]:
a = extractindex(42,n)

In [58]:
a

65

Again all these functions are fairly redundant as they are inbuilt in base python or various packages but the principles of creating your own functions are the important thing. 



What we are going to do now is create a more complex function that could have practical applications. We are going to create a classic function that calculates compound interest with a starting investment, an interest rate and a time period. 

In [61]:
def savings(start,interest,time):
    interest = 1 + (interest/100)
    current = start
    for i in range(0,time):
        current = current * interest
    
    return(current)

In [63]:
savings(1000,4,10)

1480.2442849183444

So if I start with £1000, get 4% interest a year for 10 years I will end with £1480 pounds.

In [65]:
savings(1000,4,100)

50504.94818426958

However if we wait 100 years you've got 50K. This is the power of compound interest.

However most savings accounts don't just have an initial investment and then are left for 100 years.

Most savings accounts have a regular investment into them. Add this into the function.

How about the fact that most people are paid monthly? Allow them to specify if they are putting money in monthly or annualy and then calculate it respectively.

Finally most people are investing money to save for a house or some other large purchase. Add this into the function with a goal investment and a print statement that tells them how many years they have to save for.